### DISCUSSION TITLE
2.0 migration environment variable difficulties

### DISCUSSION THREAD

**Discussion by andrevenancio** [vercel/turborepo#8432] [Help] [answered]
### Summary

As happy and supportive for everything that vercel has done for the last couple of years, I also have to say that sometimes things aren't great.

I have a monorepo running on turborepo with X amount of packages and X amounts of apps. Since the beginning it was clear to me that having a single source of environment variables on the root of the monorepo, when we setup individual projects in vercel was a bad decision. Then, through some digging and frustration I came to realise that vercel-cli was necessary so I could run `vercel env pull` which had to be done AFTER a `vercel link`. The issue with `vercel link` is that it connects to an individual app, so then I learn about `vercel link --repo` which is still an experimental feature but does what I think is expected in a monorepo.

Now, I have my monorepo connected to multiple apps on vercel and being able to pull my environment variables individually per project (which should be the default and not behind a `--repo` flag).

I'm able to run `npm run build` once my `.env` files got downloaded and I see my project build without an issue.
I push my code to github, trigger a deployment, and now all of the sudden my application fails. and it fails because there is an environment variable missing. Note that this just happens because I've upgraded to turbo v2 and did nothing else to my code.

Now, reading some more through your documentation I come to understand that a decision was made to change how environment variables work. I quote "Strict Mode for environment variables is now the default, moving from Loose Mode ([PR](https://github.com/vercel/turbo/pull/8182))".

So I think to myself: Okay, they had to change something (not really clear WHY, that linked PR says very little, I dont understand WHY the change but I'll go with it).

I try to add the `--env-mode=loose` which feels dirty. Why do I need to add extra code to run what I was running before? Why can't I just run `turbo build`? 

Then I read a bit more and I see that I need to specify which environment variables my application needs on the turbo.json file next to the build task. Why? I have individual apps with individual env variables, or even if they weren't individual they could have different values on different apps. e.g.: `GOOGLE_ANALYTICS_ID` could be one value on appA and another on appB. So why would you force me to write all environment variables explicitly on the turbo.json file? apparently for cache? 

I think you guys are crossing an invisible line of "magic" things, where nobody really knows what to do, and your documentation is mixed. On one hand your documentation is telling us to put the `.env` file on the apps and not the root, and on the next paragraph you are asking us to add the environment variables on the turbo.json root file. Bearing in mind that real apps can have multiple environment variables, a large company with multiple apps, having everything on the root turbo.json seems a bit silly to me.

Also, on your examples your turbo.json has a `globalDependencies` prop, which goes against the advice of having individual `.env` files.

So I'm just confused with how you managing to put information out there. If you rush turborepo 2 for the Ship event, that was probably a bad decision. I have a lot of faith in Vercel as a company and the engineers that are working on this tools, but you shouldn't ship features that are not documented propertly.

You should not point to PR's without giving context on why environment variables are strict.

Apologies for the rant, but I'm blocked for the second day and I'm not really sure what I'm doing wrong. my build was working before upgrading to turbo 2.0. 

And I think the invisible line is this: You are over engineering things that should be simple. Most engineers went with NextJS because it was simple. If you are making things harder to understand for the sake of having better cache you are purposely putting some developers off. And the moment you start to do that then you're opening a space for someone else to come in and simplify what you are over complicating. 

### Additional information

```js
I can jump on a call and show a repo, but I can't reproduce a public repo of the issues I'm facing.
```

**Comment by anthonyshew** [comment, marked as answer]
Hi! Responded to you [here](https://github.com/vercel/turbo/issues/8424#issuecomment-2161630955) but can answer to more that you've provided on this Discussion:

> that linked PR says very little,

I've updated the PR description now to give more context on why.

> I try to add the --env-mode=loose which feels dirty. Why do I need to add extra code to run what I was running before?

`--env-mode=loose` is a method to ease the migration path. We foresaw that it may be infeasible to sort out which environment variables you're missing on Day 1 of your migrating, so `--env-mode=loose` is available to let you update for other changes and then you can handle your environment variables later on. I recommend making your PRs small as you move through the process so you can make incremental changes.

> Bearing in mind that real apps can have multiple environment variables, a large company with multiple apps, having everything on the root turbo.json seems a bit silly to me.

This is the motivation behind [Package Configurations](https://turbo.build/repo/docs/reference/package-configurations), so it should answer directly to that concern.

> Also, on your examples your turbo.json has a globalDependencies prop, which goes against the advice of having individual .env files.

Good call! I've updated this in https://github.com/vercel/turbo/pull/8442. This was a simple oversight. Sorry for that. I'm in charge of the examples and didn't think to make this change, so thank you for your help.

> If you rush turborepo 2 for the Ship event

Turborepo 2.0 was not a part of Vercel Ship. It was shipped the week after.

---

Overall, I understand the difficulties you're having - but it seems like it's a fairly short path to getting migrated! I'd advise using `--env-mode=loose` during the first part of your migration, adding your variables to your `env` key, and then removing the `--env-mode` flag when you're ready.

**Reply by andrevenancio** [reply]
Hey Anthony. appreciate the reply. I think between this reply and the one you mention I'm starting to have a better idea of what's happening.

**Reply by andrevenancio** [reply]
Just so its clear. I probably should have re-worded my post and made the comments a bit more productive. I understand your end of the stick and you shouldn't be dealing with my own frustrations. I appreciate you and apologies for the rants.

**Reply by anthonyshew** [reply]
Don't worry about it a bit. I believe a lot in shared responsibility in software tooling. Any time you're exceedingly frustrated with our tool, there's a very, very high chance that we could have done something better on our end. Of course, even if we craft **The Perfect Doc**™️ and put it in the perfect places, it can still go unread by the folks using our tool. (Not saying that's what happened here, but we know it happens, right.)

Besides, you were plenty kind in your message. I've been outright bullied working in open-source before. I commend you for keeping it cool even when you're frustrated. 😄

You've given me plenty to go on here in ways we can get better to help out friends in the Turboverse so I'll be thinking of the kernels of truth you brought to the table here. Appreciate you taking the time to stick with us and have a conversation. Always ready to listen.

Keep shippin'!

**Reply by louisjoecodes** [reply]
Was facing the same issue yesterday, thanks a lot for starting this thread @andrevenancio and for sharing the insights @anthonyshew. Saved me quite a bit of time this morning. 

I’m bullish on Turbo and everything you folks are doing for the eco-system. It has been a delight to work with so far. Keep it up!
